### PR TITLE
Add Jacobian matrix option to Model

### LINKED ISSUE #204: Add Jacobian to Model

### PR DESCRIPTION
**Pull request by nabriis**
Closes #204

### CODE REVIEW

**Review by jakobsj — CHANGES_REQUESTED** [review]
Overall looks good, couple of small requests mainly to add one or more small examples to help convey how to specify gradient/Jacobian, as the current description while accurate may be aided by more concrete instructions such as an example.

**Review by amal-ghamdi — APPROVED** [review]
LGTM!

File: cuqi/model/_model.py
Comment (amal-ghamdi) [inline cuqi/model/_model.py:1]: @nabriis, following our discussion today,  here is a suggestion on how to update PDEModel initializer:

Change it from:
```python
    def __init__(self, PDE, range_geometry, domain_geometry):
        #....
        if not isinstance(PDE,cuqi.pde.PDE):
            raise ValueError("PDE needs to be a cuqi PDE.")
        super().__init__(self._forward_func, range_geometry, domain_geometry)
        self.pde = PDE
        if hasattr(self.pde, "gradient_wrt_parameter"):
            self._gradient_func = self.pde.gradient_wrt_parameter
```

To probably something like 
```python
    def __init__(self, PDE, range_geometry, domain_geometry):
        #....
        if not isinstance(PDE,cuqi.pde.PDE):
            raise ValueError("PDE needs to be a cuqi PDE.")
        if hasattr(self.pde, "gradient_wrt_parameter"):
            gradient = PDE.gradient_wrt_parameter
        if hasattr(self.pde, "jacobian_wrt_parameter"):
            jacobian = PDE.jacobian_wrt_parameter
        super().__init__(self._forward_func, range_geometry, domain_geometry, gradient, jacobian)
        self.pde = PDE

```

**Review by amal-ghamdi — COMMENTED** [review]
I added a comment about the PDEModel initializer

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
I am just adding request changes to keep track of the comment about the PDEModel initializer :)

File: cuqi/model/_model.py [hunk 1]
```diff
@@ -23,13 +23,21 @@ class Model(object):
...
 
-    gradient : callable function
+    gradient : callable function, optional
         The direction-Jacobian product of the forward operator Jacobian with 
         respect to the forward operator input, evaluated at a point (`wrt`).
         The signature of the gradient function should be (`direction`, `wrt`),
         where `direction` is the direction by which the Jacobian matrix is
         multiplied and `wrt` is the point at which the Jacobian is computed.
 
+    jacobian : callable function, optional
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/model/_model.py:34]: Added the example. Hope it helps

File: cuqi/model/_model.py
Comment (nabriis) [reply cuqi/model/_model.py:1]: Thanks. I added something akin to this.

File: cuqi/model/_model.py [hunk 2]
```diff
@@ -44,15 +52,26 @@ class Model(object):
...
+        # Check if only one of gradient and jacobian is given
+        if (gradient is not None) and (jacobian is not None):
+            raise TypeError("Only one of gradient and jacobian should be specified")
+        
         #Check if input is callable
         if (gradient is not None) and (callable(gradient) is not True):
             raise TypeError("Gradient needs to be callable function of some kind")
+        
+        if (jacobian is not None) and (callable(jacobian) is not True):
+            raise TypeError("Jacobian needs to be callable function of some kind")
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/model/_model.py]: Revised.

**Review by nabriis — COMMENTED** [review]
Took the comments into account @amal-ghamdi and @jakobsj. Thanks again.

**Review by jakobsj — APPROVED** [review]
Thank you for incorporating the suggestions. It looks great, in particular the detailed example is very helpful.

**Pull request merged**